### PR TITLE
Move training to Modal

### PR DESCRIPTION
**Pull request by corbt**
Use Modal for training and inference.

### CODE REVIEW

File: .tool-versions [hunk 1]
```diff
@@ -1 +1,2 @@
 nodejs 20.2.0
+python 3.9.13
```
Comment (arcticfly) [inline .tool-versions]: Nit: add newline?

File: app/Dockerfile [hunk 2]
```diff
@@ -29,7 +39,7 @@ COPY --from=deps /code/node_modules ./node_modules
 COPY --from=deps /code/app/node_modules ./app/node_modules
 COPY --from=deps /code/client-libs/typescript/node_modules ./client-libs/typescript/node_modules
 COPY . .
-RUN cd app && SKIP_ENV_VALIDATION=1 pnpm build
+# RUN cd app && SKIP_ENV_VALIDATION=1 pnpm build
```
Comment (arcticfly) [inline app/Dockerfile]: Why did you comment out the build step?

File: app/prisma/schema.prisma [hunk 3]
```diff
@@ -189,15 +188,15 @@ enum DatasetFileUploadStatus {
...
-    datasetId String   @db.Uuid
-    dataset   Dataset @relation(fields: [datasetId], references: [id], onDelete: Cascade)
-    blobName  String
-    fileName String
-    fileSize Int
-    progress Int @default(0) // Percentage
-    status   DatasetFileUploadStatus @default(PENDING)
-    uploadedAt DateTime
-    visible  Boolean @default(true)
+    datasetId    String                  @db.Uuid
```
Comment (arcticfly) [inline app/prisma/schema.prisma:191]: Do you have a formatter for prisma files?

File: app/scripts/copy-db.sh
Comment (arcticfly) [inline app/scripts/copy-db.sh:1]: Why did you remove the production debugging script?

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    max_tokens: rest.max_tokens,
+    temperature: rest.temperature ?? 0,
+    n: rest.n ?? 1,
+  };
+
+  if (input.stream) {
+    throw new Error("Streaming is not yet supported");
+  }
+
+  const resp = await fetch(`https://openpipe--${modalDeployId}.modal.run`, {
```
Comment (arcticfly) [inline app/src/modelProviders/fine-tuned/getCompletion-2.ts]: I suppose we can get rid of inference urls sometime soon.

File: app/src/modelProviders/fine-tuned/getCompletion-2.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    })),
+    usage: {
+      prompt_tokens: parsedResponse.usage.prompt_tokens,
+      completion_tokens: parsedResponse.usage.completion_tokens,
+      total_tokens: parsedResponse.usage.prompt_tokens + parsedResponse.usage.completion_tokens,
+    },
+  };
+}
+
+export const formatInputMessages = (
```
Comment (arcticfly) [inline app/src/modelProviders/fine-tuned/getCompletion-2.ts:57]: Is this all copy/pasted from getCompletion? I would prefer to define them in one place and import from there.

File: app/src/server/tasks/fineTuning/deployFineTune.task.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    data: {
+      status: "DEPLOYING",
+      deploymentStartedAt: new Date(),
+    },
+  });
+
+  const modalDeployId = `ft-${env.NODE_ENV}-${fineTuneId}`;
+
+  // Inline setupTemporaryEnvironment
+  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "deploy-"));
```
Comment (arcticfly) [inline app/src/server/tasks/fineTuning/deployFineTune.task.ts]: Should we incorporate the deploy id into the tempDir name?

File: app/src/server/tasks/fineTuning/deployFineTune.test.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        modalDeployId: "ft-local-test",
+        huggingFaceModelId: "OpenPipe/openpipe-848c83ca-df7d-4379-bb37-3b368b35f472-no-st",
+        pipelineVersion: 2,
+      },
+    });
+
+    await deployFineTune(ft.id);
+  },
+  {
+    timeout: 1000 * 60 * 60 * 20,
```
Comment (arcticfly) [inline app/src/server/tasks/fineTuning/deployFineTune.test.ts]: 20 hour timeout, I like it

Comment (arcticfly) [reply app/src/server/tasks/fineTuning/deployFineTune.test.ts]: We need to make sure to add this to our CI/CD 😆

File: app/src/tests/helpers/globalSetup.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import "./loadEnv";
+import { $ } from "execa";
+
+// await $({ stdio: "inherit" })`pnpm prisma migrate reset --force`;
```
Comment (arcticfly) [inline app/src/tests/helpers/globalSetup.ts:4]: Do we need this commented-out line?

File: trainer/pyproject.toml [hunk 9]
```diff
@@ -2,23 +2,18 @@
 name = "trainer"
 version = "0.1.0"
 description = ""
-authors = ["David Corbitt <davidlcorbitt@gmail.com>"]
```
Comment (arcticfly) [inline trainer/pyproject.toml:5]: Shots fired

File: trainer/trainer/entrypoint.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    # 8 hour timeout
+    timeout=60 * 60 * 8,
+)
+def train(fine_tune_id: str, base_url: str, api_key: str):
+    from .train import do_train
+
+    do_train(fine_tune_id, base_url, api_key)
+
+
+@stub.local_entrypoint()
```
Comment (arcticfly) [inline trainer/trainer/entrypoint.py]: Does this only get called when we're running the server on our own machines? In the production environment, what calls `train`?

**Review by arcticfly — APPROVED** [review]
Nice work! I'm very excited to try this out.

File: app/prisma/schema.prisma (hunk 3 above)
Comment (corbt) [reply app/prisma/schema.prisma:191]: Yeah just the one built into their vs code [extension](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma)

File: app/scripts/copy-db.sh
Comment (corbt) [reply app/scripts/copy-db.sh:1]: Moved it, didn't remove.

File: app/src/server/tasks/fineTuning/deployFineTune.task.ts (hunk 6 above)
Comment (corbt) [reply app/src/server/tasks/fineTuning/deployFineTune.task.ts]: Now outdated

File: trainer/pyproject.toml (hunk 9 above)
Comment (corbt) [reply trainer/pyproject.toml:5]: Lol sorry that actually wasn't intentional, at one point in this PR I was moving away from poetry and onto pip entirely so I deleted all the poetry-generated config and then ended up adding it back in. Added back. 😂

File: trainer/trainer/entrypoint.py (hunk 10 above)
Comment (corbt) [reply trainer/trainer/entrypoint.py]: Now outdated.

File: app/Dockerfile (hunk 2 above)
Comment (corbt) [reply app/Dockerfile]: Accidentally committed this, was just to make local testing faster. Added back.

File: app/src/tests/helpers/globalSetup.ts (hunk 8 above)
Comment (corbt) [reply app/src/tests/helpers/globalSetup.ts:4]: Removed
